### PR TITLE
fix: [DEL-4813]: Install proxy config through docker secret for DOCKER delegates

### PR DESCRIPTION
**Pull request by xingchi-jin**
## Describe your changes
The current way of setting proxy config is problematic when the variables contain special characters, eg. '$'
Change to use Docker Secret for proxy config setting.
Design document: https://harness.atlassian.net/wiki/spaces/~62db3e145eeecd39350a82ea/pages/21188510164/Configure+Delegate+Proxy+Credentials 

Change has been verified by building a new docker image using the new start.sh script

## Checklist
- [X] I've documented the changes in the PR description.
- [X] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>
  
- Feature build: `trigger feature-build`
- Immutable delegate `trigger publish-delegate`
</details>

<details>
  <summary>PR Check triggers</summary>
  
  You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ti0, ti1`
  
- Compile: `trigger compile`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/37944)

### CODE REVIEW

**Comment by diptiman-adak** [comment]
Quality gate Quality gate: **_NA_** 
 


            
 [Details](https://app.embold.io/organization/gh/harness/repositories/7752dbb3099d2dba3b5d37f5931a5e4c/pull-request/37944)

**Review by gaurav-jain1 — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37944)* status: 3 file(s) to review. 

___
*[`dockerization/delegate/immutable-scripts/proxy_setup.sh` line 17 at r2](https://reviewable.io/reviews/harness/harness-core/37944#-NCzsgeK4fAvsf_MRdRi:-NCzsgeK4fAvsf_MRdRj:bdklhx9) ([raw file](https://github.com/harness/harness-core/blob/29bab29e5357bb78243ee94ff92f94dcbf5f9384/dockerization/delegate/immutable-scripts/proxy_setup.sh#L17)):*
> ```Shell
> DOCKER_PROXY_SECRET_FILE="/run/secrets/proxy.config"
> if [ "$DELEGATE_TYPE" == "DOCKER" ] && [ -e "$DOCKER_PROXY_SECRET_FILE" ]; then
>   echo "docker delegate: copy proxy config mounted at ""$DOCKER_PROXY_SECRET_FILE"
> ```

%s/d/D



_Code quote:_
```Shell
d
```
___
*[`dockerization/delegate/scripts/start.sh` line 74 at r2](https://reviewable.io/reviews/harness/harness-core/37944#-NCzt6EBCrhJ01Vxqy7f:-NCzt6EBCrhJ01Vxqy7g:b-zaos) ([raw file](https://github.com/harness/harness-core/blob/29bab29e5357bb78243ee94ff92f94dcbf5f9384/dockerization/delegate/scripts/start.sh#L74)):*
> ```Shell
> fi
> 
> DOCKER_PROXY_SECRET_FILE="/run/secrets/proxy.config"
> ```

Just to confirm, the proxy.config will be similar in syntax what we expect in our delegate pod(s) right?



_Code quote:_
```Shell
/run/secrets/proxy.config
```
___
*[`dockerization/delegate/scripts/start.sh` line 76 at r2](https://reviewable.io/reviews/harness/harness-core/37944#-NCzt0p32_8Le4WhFoyw:-NCzt0p32_8Le4WhFoyx:bdklhx9) ([raw file](https://github.com/harness/harness-core/blob/29bab29e5357bb78243ee94ff92f94dcbf5f9384/dockerization/delegate/scripts/start.sh#L76)):*
> ```Shell
> DOCKER_PROXY_SECRET_FILE="/run/secrets/proxy.config"
> if [ "$DELEGATE_TYPE" == "DOCKER" ] && [ -e "$DOCKER_PROXY_SECRET_FILE" ]; then
>   echo "docker delegate: copy proxy config mounted at ""$DOCKER_PROXY_SECRET_FILE"
> ```

%s/d/D



_Code quote:_
```Shell
d
```

**Review by xingchi-jin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37944)* status: 3 file(s) to review. 1. unresolved discussions. 

___
*[`dockerization/delegate/scripts/start.sh` line 74 at r2](https://reviewable.io/reviews/harness/harness-core/37944#-NCzt6EBCrhJ01Vxqy7f:-ND-4Ji_5fdHRCMraXck:b-o9z4ec) ([raw file](https://github.com/harness/harness-core/blob/29bab29e5357bb78243ee94ff92f94dcbf5f9384/dockerization/delegate/scripts/start.sh#L74)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

Just to confirm, the proxy.config will be similar in syntax what we expect in our delegate pod(s) right?
</blockquote></details>

yes, it will be 



_Code snippet:_
```Shell
var1=""
var2=""
```
___
*[`dockerization/delegate/scripts/start.sh` line 76 at r2](https://reviewable.io/reviews/harness/harness-core/37944#-NCzt0p32_8Le4WhFoyw:-ND-Rxew4dPNa7nWhZCi:b-6p61w0) ([raw file](https://github.com/harness/harness-core/blob/29bab29e5357bb78243ee94ff92f94dcbf5f9384/dockerization/delegate/scripts/start.sh#L76)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

%s/d/D
</blockquote></details>

can you elaborate ?

File: 400-rest/src/main/resources/delegatetemplates/harness-ng-delegate-docker-compose.yaml.ftl [hunk 1]
```diff
@@ -30,3 +30,5 @@ services:
       - DELEGATE_TAGS=${delegateTags}
       - PROXY_MANAGER=true
       - INIT_SCRIPT=echo "Docker delegate init script executed."
+# If proxy setting is required, Please refer to
```
Comment (xingchi-jin) [inline 400-rest/src/main/resources/delegatetemplates/harness-ng-delegate-docker-compose.yaml.ftl:33]: feel free to fix me for any better language

**Review by gaurav-jain1 — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37944)* status: 5 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/resources/delegatetemplates/start.sh.ftl` line 45 at r4](https://reviewable.io/reviews/harness/harness-core/37944#-ND2MKhkAUjeqkQFeWei:-ND2MKhkAUjeqkQFeWej:biv2y0d) ([raw file](https://github.com/harness/harness-core/blob/f48b62f50f94ba6473a6af83ac6fde5ff742987f/400-rest/src/main/resources/delegatetemplates/start.sh.ftl#L45)):*
> ```FreeMarker
> DOCKER_PROXY_SECRET_FILE="/run/secrets/proxy.config"
> if [ "$DELEGATE_TYPE" == "DOCKER" ] && [ -e "$DOCKER_PROXY_SECRET_FILE" ]; then
>   echo "docker delegate: copy proxy config mounted at ""$DOCKER_PROXY_SECRET_FILE"
> ```

I don't think we need changes here, as the we already have copied the proxy file during container start as part of your changes in `start.sh`



_Code quote:_
```FreeMarker
DOCKER_PROXY_SECRET_FILE="/run/secrets/proxy.config"
if [ "$DELEGATE_TYPE" == "DOCKER" ] && [ -e "$DOCKER_PROXY_SECRET_FILE" ]; then
  echo "docker delegate: copy proxy config mounted at ""$DOCKER_PROXY_SECRET_FILE"
  cp "$DOCKER_PROXY_SECRET_FILE" 'proxy.config'
fi
```
___
*[`dockerization/delegate/scripts/start.sh` line 76 at r2](https://reviewable.io/reviews/harness/harness-core/37944#-NCzt0p32_8Le4WhFoyw:-ND2K71a2zvv_SOr0aqw:b-9dstlm) ([raw file](https://github.com/harness/harness-core/blob/29bab29e5357bb78243ee94ff92f94dcbf5f9384/dockerization/delegate/scripts/start.sh#L76)):*
<details><summary><i>Previously, xingchi-jin (Xingchi Jin) wrote…</i></summary><blockquote>

can you elaborate ? 
</blockquote></details>

make first letter to uppercase `D`



_Code quote:_
```Shell
docker
```

**Review by xingchi-jin — COMMENTED** [review]
resolved comments 

> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37944#-:-ND4S_lR7K1NF9fq6xLn:bi5plxl)* status: 5 file(s) to review. 1 change requests. 1. unresolved discussions. 0 of 1 approvals obtained. 

___
*[`400-rest/src/main/resources/delegatetemplates/start.sh.ftl` line 45 at r4](https://reviewable.io/reviews/harness/harness-core/37944#-ND2MKhkAUjeqkQFeWei:-ND4SXL7D7-INqEQ19qZ:b-896fix) ([raw file](https://github.com/harness/harness-core/blob/f48b62f50f94ba6473a6af83ac6fde5ff742987f/400-rest/src/main/resources/delegatetemplates/start.sh.ftl#L45)):*
<details><summary><i>Previously, gaurav-jain1 (Gaurav Jain) wrote…</i></summary><blockquote>

I don't think we need changes here, as the we already have copied the proxy file during container start as part of your changes in `start.sh`
</blockquote></details>

Done.

**Review by raghuAtWings — DISMISSED** [review]
> Reviewed 1 of 2 files at r3, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37944)* status: 1 change requests. 1. unresolved discussions.

**Comment by xingchi-jin** [comment]
trigger messagecheck

**Comment by xingchi-jin** [comment]
trigger ti1

**Review by gaurav-jain1 — DISMISSED** [review]
> Reviewed 1 of 2 files at r3, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37944)* status: :shipit: complete! ✓

**Comment by xingchi-jin** [comment]
trigger ti1

**Review by raghuAtWings — APPROVED** [review]
> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/37944)* status: :shipit: complete! ✓

**Pull request merged**